services/app/src/components/views/quest/QuestEnd.tsx

```diff
@@ -46,48 +216,54 @@ export default class QuestEnd extends React.Component<Props, {}> {
+    } else {
+      return `Primary Issue: ${this.state.primaryIssue || 'not given'}\nQualifier: ${this.state.issueQualifier || 'not given'}\nDetails: ${this.state.text}`;
+    }
+  }
+
+  private validateAndSubmit() {
+    if (this.isBadRating() && !this.state.primaryIssue) {
+      this.setState({needsIssue: true});
+      return;
+    } else {
+      this.setState({needsIssue: false});
+    }
+    this.props.onSubmit(this.props.quest, this.props.settings, this.props.user, this.state.anonymous, this.formatFeedback(), this.state.rating);
+  }
+
+  public render() {
+    const rated = this.state.rating !== null && (this.state.rating > 0);
     return (
       <Card title={this.props.quest.details.title}>
         <p>We hope you enjoyed <i>{this.props.quest.details.title}</i> by {this.props.quest.details.author}!</p>
         <p>Rate this quest:</p>
-        <StarRating hintText={true} value={this.state.rating || 0} onChange={(rating: number) => { this.setState({rating}); }}></StarRating>
-        {rated &&
-          <div>
-            <TextField
-              className="textfield"
-              fullWidth={true}
-              label={<span>Write a short review</span>}
-              multiline={true}
-              margin="normal"
-              onChange={(e: any) => this.setState({text: e.target.value})}
-              onFocus={(e: any) => e.target.scrollIntoView()}
-              value={this.state.text}
-            />
-            <FormControlLabel
-              control={
-                  <Checkbox
-                    className="anonymous_feedback"
-                    checked={this.state.anonymous}
-                    onChange={() => { this.setState({anonymous: !this.state.anonymous}); }}/>
-              }
-              label="Give feedback anonymously"/>
-          </div>
-        }
-        Tip the author{this.state.rating && ' and submit your review:'}
-        <div className={'tipAmounts ' + (checkoutError === null ? '' : 'checkoutDisabled')}>
-          {tips}
-        </div>
-        <Button onClick={() => this.props.onSubmit(this.props.quest, this.props.settings, this.props.user, this.state.anonymous, this.state.text, this.state.rating)}>
+        <StarRating id="starrating" hintText={true} value={this.state.rating || 0} onChange={(rating: number) => { this.setState({rating}); }}></StarRating>
+        {this.renderFeedbackForm()}
+        {this.maybeRenderTipping()}
+        <Button id="submit" onClick={() => this.validateAndSubmit()}>
           {rated ? 'Submit' : 'Return home'}
         </Button>
-        {window.plugins && window.plugins.socialsharing &&
-          <Button onClick={() => this.props.onShare(this.props.quest)}><img className="inline_icon" src="images/share_small.svg"/> Share your adventure</Button>
+        {this.props.showSharing &&
+          <Button id="shareButton" onClick={() => this.props.onShare(this.props.quest)}>
```

Comment:
Does this mean we can get rid of the declare var window?

Reply:
Done.